service/service.go

```diff
@@ -158,74 +123,43 @@ func (s *Service) Register(
-	defer span.End()
+	reqCtx := pipeline.NewRequestContext(
+		streamSrv.Context(),
+		request,
+		s.tracer,
+	)
+	reqCtx.StartSpan("substreams_request")
+	defer reqCtx.EndSpan()
 
-	logger := logging.Logger(ctx, s.logger)
-	hostname, err := os.Hostname()
-	if err != nil {
-		logger.Warn("cannot find hostname, using 'unknown'", zap.Error(err))
-		hostname = "unknown host"
-	}
-	if os.Getenv("SUBSTREAMS_SEND_HOSTNAME") == "true" {
-		md := metadata.New(map[string]string{"host": hostname})
-		err = streamSrv.SetHeader(md)
-		if err != nil {
-			logger.Warn("cannot send header metadata", zap.Error(err))
-		}
-	}
-	span.SetAttributes(attribute.String("hostname", hostname))
-
-	logger.Info("serving blocks")
-	if request.StartBlockNum < 0 {
-		// TODO(abourget) start block resolving is an art, it should be handled here
-		err := fmt.Errorf("invalid negative startblock (not handled in substreams): %d", request.StartBlockNum)
-		span.SetStatus(otelcode.Error, err.Error())
-		return err
-	}
+	reqCtx.SetAttributes(attribute.StringSlice("module_outputs", request.OutputModules))
+	reqCtx.SetLogger(logging.Logger(streamSrv.Context(), s.logger))
 
-	if request.Modules == nil {
-		err := status.Error(codes.InvalidArgument, "no modules found in request")
-		span.SetStatus(otelcode.Error, err.Error())
-		return err
-	}
+	hostname := updateStreamHeadersHostname(streamSrv, reqCtx.Logger())
+	reqCtx.SetAttributes(attribute.String("hostname", hostname))
 
-	if err := manifest.ValidateModules(request.Modules); err != nil {
-		err := status.Error(codes.InvalidArgument, fmt.Sprintf("modules validation failed: %s", err))
-		span.SetStatus(otelcode.Error, err.Error())
-		return err
+	if grpcError := s.blocks(reqCtx, streamSrv); grpcError != nil {
+		reqCtx.SetErr(grpcError.cause())
+		return grpcError.rpcErr()
 	}
+	reqCtx.SetOk("")
+	return nil
+}
 
-	if err := pbsubstreams.ValidateRequest(request); err != nil {
-		err := status.Error(codes.InvalidArgument, fmt.Sprintf("validate request: %s", err))
-		span.SetStatus(otelcode.Error, err.Error())
-		return err
-	}
+func (s *Service) blocks(request *pipeline.RequestContext, streamSrv pbsubstreams.Stream_BlocksServer) GRPCError {
+	request.Logger().Info("validating request")
+	// TODO find a better name
 
-	graph, err := manifest.NewModuleGraph(request.Modules.Modules)
+	graph, err := s.requestValidator.Validate(request.Request())
 	if err != nil {
-		err := fmt.Errorf("creating module graph %w", err)
-		span.SetStatus(otelcode.Error, err.Error())
-		return err
-	}
-
-	sources := graph.GetSources()
-	for _, source := range sources {
-		if source != s.blockType && source != "sf.substreams.v1.Clock" {
-			err := fmt.Errorf(`input source %q not supported, only %q and "sf.substreams.v1.Clock" are valid`, source, s.blockType)
-			span.SetStatus(otelcode.Error, err.Error())
-			return err
-		}
+		return NewBasicErr(status.Error(codes.InvalidArgument, err.Error()), err)
 	}
```

Comment:
have you guys checked what's in `derr` ? We have things to wrap/unwrap gRPC errors, etc..